File: java/src/main/java/io/github/matheuslm7/brvalidator/ie/Pr.java
package io.github.matheuslm7.brvalidator.ie;

import io.github.matheuslm7.brvalidator.internal.Shared;
import java.util.regex.Pattern;

/**
 * PR: verified against the official SEFAZ-PR "Roteiro de Crítica da Inscrição Estadual"
 * (sintegra.gov.br/Cad_Estados/cad_PR.html), including its fully worked example
 * (123.45678-50), and cross-checked against the reference Visual Basic routine published on the
 * same page. Format: 8 digits + 2 check digits.
 */
final class Pr {

  private static final Pattern ALLOWED_CHARS = Pattern.compile("^[\\d.\\-\\s]+$");
  private static final int[] FIRST_DIGIT_WEIGHTS = {3, 2, 7, 6, 5, 4, 3, 2};
  private static final int[] SECOND_DIGIT_WEIGHTS = {4, 3, 2, 7, 6, 5, 4, 3, 2};

  private Pr() {}

  static String normalize(String value) {
    return Shared.removeNonDigits(value);
  }

  static boolean isValid(String value) {
    if (!ALLOWED_CHARS.matcher(value).matches()) {
      return false;
    }

    String v = normalize(value);

    if (v.length() != 10) {
      return false;
    }

    String base = v.substring(0, 8);
    int firstDigit = Mod11.mod11CheckDigit(Mod11.weightedSum(base, FIRST_DIGIT_WEIGHTS));

    if (firstDigit != (v.charAt(8) - '0')) {
      return false;
    }

    int secondDigit =
        Mod11.mod11CheckDigit(Mod11.weightedSum(base + firstDigit, SECOND_DIGIT_WEIGHTS));

    return secondDigit == (v.charAt(9) - '0');
  }

  static String format(String value) {
    String v = normalize(value);

    if (v.length() != 10) {
      return value;
    }

    return v.substring(0, 3) + "." + v.substring(3, 8) + "-" + v.substring(8, 10);
  }
}

File: src/ie/states/pr.ts
import { removeNonDigits } from "../../shared/normalize.js";
import { weightedSum, mod11CheckDigit } from "../../shared/mod11.js";

// Verified against the official SEFAZ-PR "Roteiro de Crítica da Inscrição
// Estadual" (sintegra.gov.br/Cad_Estados/cad_PR.html), including its fully
// worked example (123.45678-50), and cross-checked against the reference
// Visual Basic routine published on the same page. Format: 8 digits + 2
// check digits.
const FIRST_DIGIT_WEIGHTS = [3, 2, 7, 6, 5, 4, 3, 2];
const SECOND_DIGIT_WEIGHTS = [4, 3, 2, 7, 6, 5, 4, 3, 2];

function normalize(value: string): string {
  return removeNonDigits(value);
}

function isValid(value: string): boolean {
  if (!/^[\d.\-\s]+$/.test(value)) {
    return false;
  }

  const ie = normalize(value);

  if (ie.length !== 10) {
    return false;
  }

  const base = ie.slice(0, 8);
  const firstDigit = mod11CheckDigit(weightedSum(base, FIRST_DIGIT_WEIGHTS));

  if (firstDigit !== Number(ie[8])) {
    return false;
  }

  const secondDigit = mod11CheckDigit(weightedSum(base + firstDigit, SECOND_DIGIT_WEIGHTS));

  return secondDigit === Number(ie[9]);
}

function format(value: string): string {
  const ie = normalize(value);

  if (ie.length !== 10) {
    return value;
  }

  return ie.replace(/^(\d{3})(\d{5})(\d{2})$/, "$1.$2-$3");
}

export const PR = {
  isValid,
  normalize,
  format,
};
